### PR TITLE
Matching algorithm for procedures and reports

### PR DESCRIPTION
**Pull request by RadmirGari**
_[Release PRs:]_

Issues:

- https://linear.app/metriport/issue/ENG-1006

### Dependencies
None

### Description
Add a matching algorithm that matches reports and procedures based off of:
1.  Date
2. Codes
3. Identifier Values

### Testing
- Local
  - [x] Test on different XMLs to make sure imaging reports are linked [(IMG IN THREAD)](https://metriport.slack.com/archives/C0616FCPAKZ/p1757449808189129?thread_ts=1756944508.559769&cid=C0616FCPAKZ)
  - [x] Run Unit tests
- Production
  - [ ] Reconvert single patient to make sure Alexey's changes are working with these changes.

## Summary by CodeRabbit

- **New Features**
  - Procedures are now automatically linked to related Diagnostic Reports during normalization using identifier/code token matching and near-time (2‑hour) date alignment. Existing references are preserved, duplicates avoided, and the first suitable match is preferred; final bundles include these relationships.

- **Tests**
  - Added comprehensive tests for date matching and procedure–report linking, covering identifier normalization, multiple-match scenarios, invalid inputs, edge cases, identifier formats, trailing-carets, and window-based comparisons.

### CODE REVIEW

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  const raw = [
+    dr.effectiveDateTime,
+    dr.effectivePeriod?.start,
+    dr.effectivePeriod?.end,
+    dr.issued,
+  ].filter(Boolean) as string[];
+  return [...new Set(raw)];
+}
+
+function keysForDiagnosticReport(dr: DiagnosticReport): string[] {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: rename to indicate an action. Something like `getKeys..` or `buildKeys..`

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  const idVals = identifierValueTokens(dr);
+  const codes = codeTokensFromCode(dr.code);
+
+  const out: string[] = [];
+  for (const v of idVals) out.push(`${v}`);
+  for (const c of codes) out.push(`${c}`);
+  return dedupe(out);
+}
+
+function keysForProcedure(p: Procedure): { dates: string[]; pKeys: string[] } {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: ditto: func name

**Review by RamilGaripov — CHANGES_REQUESTED** [review]
plz update all func names to indicate actions

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 3]
```diff
@@ -0,0 +1,638 @@
+import { buildDayjs } from "@metriport/shared/common/date";
+import {
+  linkProceduresToDiagnosticReports,
+  matchDates,
+  SIZE_OF_WINDOW,
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: fixed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    .split("|")
+    .map(s => s.trim())
+    .filter(Boolean);
+  return parts[parts.length - 1] ?? "";
+}
+
+function splitValueByDash(value: string): string {
+  const [head = "", tail = ""] = (value ?? "").split("-", 2);
+  return head && tail && /^\d+$/.test(head) && /^\d+$/.test(tail) ? head : value ?? "";
+}
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: @RamilGaripov @Orta21 Thoughts on this?
Lets say we get 2 LOINC codes:
```
123-4
123-6
```
Right now we will truncate them both to just `123`
and consider them a match

 rabbit is basically saying that we shouldn't do this to get better matches. 
 But from my understanding 
 `123` means some category
 and the -X means some more specific thing, so it's possible that the procedure say could be just
 `123` and the diagnostic report
` 123-6` 
applying the change would mean `123-4` and `123-6` no longer match.
But if we are going for more accuracy then applying the change makes sense

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+        effectiveDateTime: matchingDate,
+      });
+
+      const result = linkProceduresToDiagnosticReports([procedure], [diagnosticReport]);
+
+      expect(result.procedures[0]?.report).toBeDefined();
+      expect(result.procedures[0]?.report?.[0]?.reference).toBe(
+        `DiagnosticReport/${diagnosticReport.id}`
+      );
+    });
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: agreed fixed.

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  ) as string[];
+  return dedupe(raw);
+}
+
+function identifierValueTokens(dr: Resource): string[] {
+  const BAD = new Set(["UNK", "UNKNOWN", "UNSPECIFIED", "NA", "N/A", "NONE", "NO_CODE"]);
+  const out: string[] = [];
+  if (!("identifier" in dr)) {
+    throw new Error("No identifier in this resource");
+  }
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: agreed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: why would we strip off parts of medical codes? keep the loinc codes intact

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
+import {
+  CodeableConcept,
+  Coding,
+  DiagnosticReport,
+  Procedure,
+  Reference,
+  Resource,
+} from "@medplum/fhirtypes";
+
+const HOURS_FOR_WINDOW = 2;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: use `dayjs.duration()` to define const related to time

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Okay will remove all splits by dash

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+export const SIZE_OF_WINDOW = HOURS_FOR_WINDOW * 60 * 60 * 1000;
+
+export function linkProceduresToDiagnosticReports(
+  procedures: Procedure[],
+  reports: DiagnosticReport[]
+): { procedures: Procedure[]; reports: DiagnosticReport[] } {
+  const drByKey = new Map<string, DiagnosticReport[]>();
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: `drMap` is good

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+  const codes = codeTokensFromCode(dr.code);
+
+  const out: string[] = [];
+  for (const v of idVals) out.push(`${v}`);
+  for (const c of codes) out.push(`${c}`);
+  return dedupe(out);
+}
+
+function getKeysAndDatesForProcedure(p: Procedure): { dates: string[]; pKeys: string[] } {
+  const dates = dateCandidatesFromProcedure(p);
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: we already have this - `getPerformedDateFromResource`, could reuse here

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:1]: update the logic around date comparisons using `buildDayJs` and its helpers. Refer to this func - `isEligibleForRedownload`

**Review by RamilGaripov — CHANGES_REQUESTED** [review]

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:1]: Done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 9 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 8 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 7 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 2 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 1 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+function getIdentifierValueTokens(dr: Resource): string[] {
+  const BAD = new Set(["UNK", "UNKNOWN", "UNSPECIFIED", "NA", "N/A", "NONE", "NO_CODE"]);
+  const out: string[] = [];
+  if (!("identifier" in dr)) {
+    return [];
+  }
+
+  const identifiers = Array.isArray(dr.identifier) ? dr.identifier : [dr.identifier];
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: we have a toArray() func for this

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+  for (const id of identifiers) {
+    if (!id) continue;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: redundant

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    const value = (id.value ?? "").toString().trim();
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: `(id.value ?? "").toString()` is also redundant

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 10 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+    if (BAD.has(valueByPipe.toUpperCase())) continue;
+    if (/^(?:urn:|oid:|https?:\/\/)/i.test(valueByPipe)) continue;
+    const noTrailingCaret = removeTrailingCaret(valueByPipe);
+    out.push(noTrailingCaret);
+  }
+  return dedupe(out);
+}
+
+function getSplitValueByPipe(value: string): string {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: what's this function for? why do we need to split the the id by pipe? if there's a good reason for us to do this, put that into a tsdoc, so we know in the future

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+function getCodeTokensFromCode(code?: CodeableConcept): string[] {
+  const BAD = new Set(["UNK", "UNKNOWN", "UNSPECIFIED", "NA", "N/A", "NONE", "NO_CODE"]);
+  const codings = (code?.coding ?? []).filter((c): c is Coding => !!c?.code);
+
+  const out: string[] = [];
+  for (const c of codings) {
+    const token = (c.code ?? "").toString().trim().toUpperCase();
+    if (!token || BAD.has(token)) continue;
+    if (/^(?:urn:|oid:|https?:\/\/)/i.test(token)) continue;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: put this regex into a const outside the func

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+  }
+  return dedupe(out);
+}
+
+function dedupe<T>(arr: T[]): T[] {
+  return [...new Set(arr)];
+}
+
+function removeTrailingCaret(s: string): string {
+  if (s.charAt(s.length - 1) === "^") {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: can simplify with `s.endsWith("^")`. Also need a TSDoc to explain why we do this

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    return s.slice(0, s.length - 1);
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: i believe you can simplify to `.slice(0, -1); `. but double check plz

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+      const prior = drMap.get(k) ?? [];
+      prior.push(dr);
+      drMap.set(k, prior);
+    }
+  }
+
+  for (const proc of procedures) {
+    const { dates, pKeys } = getKeysAndDatesForProcedure(proc);
+    for (const key of pKeys) {
+      const hits = key ? drMap.get(key) : undefined;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: why are you checking if what youre iterating over exists `key ?`? it's redundant

Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 16 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 15 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done
We need this because procedure's identifiers sometimes end with a caret while diagnostic reports identifiers NEVER (from what I've seen) end with a caret. resulting in a miss-match when it shouldn't be.

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 14 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 11 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 12 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 13 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed it i think it is unneccessary

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+        if (!existing.some(r => r.reference === ref.reference)) {
+          proc.report = [...existing, ref];
+        }
+        break;
+      }
+    }
+  }
+
+  return { procedures, reports };
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: just realized 2 things here:
1. We don't need to return the reports cuz they aren't being updated in any way
2. This function is mutative. We try to avoid that unless it's absolutely needed for performance reasons:
  - I'm not feeling 100% confident approving this, so I'd want input from @leite08 or @thomasyopes 
  - If we do keep it mutative, the function name has to indicate that. Lookup `dangerously...` for other examples in the codebase

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+    for (const k of keys) {
+      if (!k) continue;
+      const prior = drMap.get(k) ?? [];
+      prior.push(dr);
+      drMap.set(k, prior);
+    }
+  }
+
+  for (const proc of procedures) {
+    const { dates, pKeys } = getKeysAndDatesForProcedure(proc);
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: what's `pKeys`? this just adds confusion. Just call it keys

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    for (const key of pKeys) {
+      const hits = drMap.get(key);
+      if (!hits) continue;
+      for (const hit of hits) {
+        if (!hit?.id) continue;
+        const hitDates = getDateFromDiagnosticReport(hit);
+        if (!doAnyDatesMatchThroughWindow(dates, hitDates)) {
+          continue;
+        }
+        const ref: Reference<DiagnosticReport> = { reference: `DiagnosticReport/${hit.id}` };
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: i think we have a function to build references - `buildResourceReference`

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+      drMap.set(k, prior);
+    }
+  }
+
+  for (const proc of procedures) {
+    const { dates, pKeys } = getKeysAndDatesForProcedure(proc);
+    for (const key of pKeys) {
+      const hits = drMap.get(key);
+      if (!hits) continue;
+      for (const hit of hits) {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: hit, hits.. dont like the naming here. a `hit` is a `diagReport`, right?

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+          continue;
+        }
+        const ref: Reference<DiagnosticReport> = { reference: `DiagnosticReport/${hit.id}` };
+        const existing: Reference<DiagnosticReport>[] = (proc.report ??
+          []) as Reference<DiagnosticReport>[];
+
+        if (!existing.some(r => r.reference === ref.reference)) {
+          proc.report = [...existing, ref];
+        }
+        break;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: why the break? are you saying there's only ever 1 match? If yes, why? if not, why do we choose the first one that we encounter in this algo? This seems non-deterministic. 

Should we add some more logic that if there's multiple matches based on the 2-hour window or something, we choose the one that's closest? I don't wanna overcomplicate this too much, but I also want every decision of this logic to make sense. 

Also, while we're at it, leave a comment for why we're using a 2-hour window.

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+  for (const c of codes) out.push(`${c}`);
+  const dedupedOut = dedupe(out);
+  const dedupedDates = dedupe(dates);
+  return { dates: dedupedDates, pKeys: dedupedOut };
+}
+
+function getDateFromDiagnosticReport(dr: DiagnosticReport): string[] {
+  const dates: string[] = [];
+
+  // Use the shared function for effectiveDateTime and effectivePeriod
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: did Cursor generate this? this comment is redundant

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+  const dateFromResource = getDateFromResource(dr, "datetime");
+  if (dateFromResource) {
+    dates.push(dateFromResource);
+  }
+
+  // Also check the issued field which is specific to DiagnosticReport
+  if (dr.issued) {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:88]: what's this "issued" field? I don't recall us using that..

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+  }
+  return dedupe(out);
+}
+
+function getCodeTokensFromCode(code?: CodeableConcept): string[] {
+  const codings = (code?.coding ?? []).filter((c): c is Coding => !!c?.code);
+
+  const out: string[] = [];
+  for (const c of codings) {
+    const token = (c.code ?? "").toString().trim();
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: `(c.code ?? "").toString()` redundant

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+ * Removes the trailing caret from a string.
+ * This is needed because the Procedures identifier values sometimes are stored with a caret at the end.
+ * But the DiagnosticReport identifier values are stored without the caret. Resulting in a mismatch.
+ * @param s The string to remove the trailing caret from
+ * @returns The string with the trailing caret removed
+ */
+function removeTrailingCaret(s: string): string {
+  if (s.endsWith("^")) {
+    return s.slice(0, -1);
+  }
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:152]: hypernit: don't need 3 lines for this code - 1 is enough

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 19 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: short for procedure keys. Just gonna call it procedureKeys

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 18 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Ill msg raf

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 24 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:88]: sent dm of it

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 23 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 25 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 22 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: removed the break, I agree with multiple reports being added

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 26 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts:152]: Ima keep it as 3 because it looks nicer that way :p

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 20 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Theres a build reference function from fhir im importing instead now

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 21 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: changed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+    const { dates, procedureKeys } = getKeysAndDatesForProcedure(proc);
+    for (const key of procedureKeys) {
+      const drs = drMap.get(key);
+      if (!drs) continue;
+      for (const dr of drs) {
+        const drDates = getDateFromDiagnosticReport(dr);
+        if (!doAnyDatesMatchThroughWindow(dates, drDates)) {
+          continue;
+        }
+        const ref: Reference<DiagnosticReport> = createReference(dr);
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: oh shit, this is a medplum's function? have you made sure this does what we expect it to do?

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+        const existing: Reference<DiagnosticReport>[] = (proc.report ??
+          []) as Reference<DiagnosticReport>[];
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Try to avoid casting as much as possible. Casting is the way to turn compiler-time errors into runtime errors, so be wary of doing that. 

In this case, can just bypass specifying types entirely:

Screenshot 2025-09-10 at 8 24 20 AM

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  return [...new Set(dates)];
+}
+
+export function doAnyDatesMatchThroughWindow(a: string[] = [], b: string[] = []): boolean {
+  if (a.length === 0 || b.length === 0) return false;
+
+  const aDates = a.map(dateStr => buildDayjs(dateStr)).filter(dayjs => dayjs.isValid());
+
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: superhypernit: styling - no line between these two, it reads cleaner :)

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 30]
```diff
@@ -0,0 +1,175 @@
...
+  const bDates = b.map(dateStr => buildDayjs(dateStr)).filter(dayjs => dayjs.isValid());
+
+  if (aDates.length === 0 || bDates.length === 0) return false;
+
+  for (const dateA of aDates) {
+    for (const dateB of bDates) {
+      const diffInMs = Math.abs(dateA.diff(dateB, "milliseconds"));
+      if (diffInMs <= SIZE_OF_WINDOW) return true;
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: nit: we define constants as duration objects, and then in code, use `SIZE_OF_WINDOW.asMilliseconds()`. 
Also, the name for this constant is super unclear. What is a "window"?

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+    }
+  }
+
+  return false;
+}
+
+function getIdentifierValueTokens(dr: Resource): string[] {
+  const out: string[] = [];
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: nit: bad const name. what's `out`?

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 32]
```diff
@@ -0,0 +1,775 @@
...
+  SIZE_OF_WINDOW,
+} from "../link-procedures-to-reports";
+import { makeProcedure } from "../../../../fhir-to-cda/cda-templates/components/__tests__/make-procedure";
+import { makeDiagnosticReport } from "../../../../fhir-to-cda/cda-templates/components/__tests__/make-diagnostic-report";
+
+describe("dangerouslyLinkProceduresToDiagnosticReports", () => {
+  const baseMs = Date.now();
+  const DATE_TO_MATCH = buildDayjs(baseMs).toISOString();
+
+  const defaultIdentifier = [
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: agreed

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 27 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: It adds on an extra field "display". I used our own just in case.

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 28 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 30 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Left a comment and also changes to .asMilliseconds

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 31 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: changed to just ids

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 33]
```diff
@@ -0,0 +1,165 @@
...
+    if (!token || isUselessDisplay(token)) continue;
+    if (URI_PATTERN.test(token)) continue;
+    out.push(token.toUpperCase());
+  }
+  return dedupe(out);
+}
+
+function dedupe<T>(arr: T[]): T[] {
+  return [...new Set(arr)];
+}
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: nit: we could just initiate accumulators as Sets and return as arrays. Then we don't need this func at all

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts:1]: update the tests so each scenario makes sense and only uses what's relevant

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 34]
```diff
@@ -0,0 +1,476 @@
...
+        },
+        performedDateTime: DATE_TO_MATCH,
+      });
+
+      const diagnosticReport = makeDiagnosticReport({
+        code: {
+          coding: [
+            {
+              code: sharedCode,
+              system: "http://www.ama-assn.org/go/cpt",
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: nit: the DR will NEVER have a CPT code afaik. We hope to match on LOINC codes, right?

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 35]
```diff
@@ -0,0 +1,476 @@
...
+
+      expect(result[0]?.report).toBeDefined();
+      expect(result[0]?.report?.[0]?.reference).toBe(`DiagnosticReport/${diagnosticReport.id}`);
+    });
+
+    it("should filter out all known useless display values", () => {
+      const matchingDate = buildDayjs(baseMs + THRESHOLD.asMilliseconds() / 2).toISOString();
+      const uselessValues = ["unknown", "unk", "no known", "no data available"];
+
+      for (const uselessValue of uselessValues) {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: i dont think we should iterate. Just pick one of the useless codes to test with

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts [hunk 36]
```diff
@@ -0,0 +1,476 @@
...
+        });
+
+        const result = linkProceduresToDiagnosticReports([procedure], [diagnosticReport]);
+
+        expect(result).toBeDefined();
+        expect(result[0]?.report).toBeUndefined();
+      }
+    });
+
+    it("should remove trailing carets from identifier values", () => {
```
Comment (RamilGaripov) [inline packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: more like "should match despite the trailing caret in procedure identifier value"

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 30 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 29 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts (hunk 33 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: done

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts:1]: done

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts (hunk 34 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: Switched to LOINC_URL

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts (hunk 35 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: done

File: packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts (hunk 36 above)
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/__tests__/normalize-matching.test.ts]: Done

File: packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts [hunk 37]
```diff
@@ -0,0 +1,154 @@
...
+  if (a.length === 0 || b.length === 0) return false;
+  const aDates = a.map(dateStr => buildDayjs(dateStr)).filter(dayjs => dayjs.isValid());
+  const bDates = b.map(dateStr => buildDayjs(dateStr)).filter(dayjs => dayjs.isValid());
+  if (aDates.length === 0 || bDates.length === 0) return false;
+
+  for (const dateA of aDates) {
+    for (const dateB of bDates) {
+      const diffInMs = Math.abs(dateA.diff(dateB, "milliseconds"));
+      if (diffInMs <= THRESHOLD.asMilliseconds()) return true;
+    }
```
[the remark this answers is not in the document]
Comment (RadmirGari) [reply packages/core/src/external/fhir/normalization/link-procedures-to-reports.ts]: Agreed done

**Review by RamilGaripov — APPROVED** [review]
:shipit: :shipit: :shipit:
